File: conformance/adapters/python/src/conformance_adapter/runner.py
# Copyright 2025 The MCP Interceptors Authors. All rights reserved.
# Use of this source code is governed by an Apache-2.0
# license that can be found in the LICENSE file.

"""The Python conformance RUNNER: replays the SHARED fixtures against an
adapter and diffs canonical JSON — the Python realization of the algorithm in
`conformance/ADAPTER.md`, structurally mirroring `conformance/src/runner.ts`.

No expectation logic is ported from TypeScript: the fixtures are the oracle.
The runner only implements the comparison model (identical in every language):

  1. scrub volatile keys (``fixture.volatile``) recursively from the ACTUAL;
  2. JCS-canonicalize the scrubbed actual (``canonical.py``, byte-equal to TS);
  3. string-compare against the fixture's precomputed ``canonical``.

Behavior expectations additionally check decision equality and
forbids/requires substrings over the canonical final payload.
"""

from __future__ import annotations

import math
from collections.abc import Awaitable, Callable
from contextlib import AbstractAsyncContextManager
from dataclasses import dataclass
from typing import Any, Literal, Protocol

from conformance_adapter.canonical import canonicalize
from conformance_adapter.fixtures import (
    ChainStep,
    DecisionExpectation,
    ErrorExpectation,
    Fixture,
    FixtureInterceptor,
    FixtureStep,
    InvokeParams,
    InvokeStep,
    ListStep,
    Phase,
)

# ── the adapter contract (ADAPTER.md, four functions) ────────────────────────


@dataclass(frozen=True, kw_only=True)
class InvokeOutcome:
    """``ok`` with the wire result, or the JSON-RPC error code."""

    ok: bool
    result: Any = None
    error_code: int | None = None


@dataclass(frozen=True, kw_only=True)
class ChainOutcome:
    #: ``allow`` iff the chain completed; ``deny`` iff it was blocked/aborted.
    decision: Literal["allow", "deny"]
    #: The payload after all applied mutations (meaningful when allowed).
    final_payload: Any


class AdapterSession(Protocol):
    def list(self, event: str | None) -> Awaitable[Any]:
        """Wire JSON of ``interceptors/list`` (event filter or None)."""
        ...

    def invoke(self, params: InvokeParams) -> Awaitable[InvokeOutcome]:
        """Wire JSON of ``interceptor/invoke``, or the JSON-RPC error code."""
        ...

    def chain(
        self, event: str, phase: Phase, payload: Any, session_id: str | None
    ) -> Awaitable[ChainOutcome]:
        """One chain execution across ALL of the session's interceptors."""
        ...


class Adapter(Protocol):
    @property
    def name(self) -> str: ...

    def create_session(
        self, interceptors: tuple[FixtureInterceptor, ...]
    ) -> AbstractAsyncContextManager[AdapterSession]:
        """Fresh implementation state with these interceptors registered."""
        ...


# ── scrubbing + canonical comparison ─────────────────────────────────────────


def scrub(value: Any, volatile: tuple[str, ...]) -> Any:
    """Recursively drop volatile keys (timings, impl-specific info) from actuals."""
    if isinstance(value, list):
        return [scrub(v, volatile) for v in value]
    if isinstance(value, dict):
        return {k: scrub(v, volatile) for k, v in value.items() if k not in volatile}
    return value


@dataclass(frozen=True, kw_only=True)
class _Comparison:
    passed: bool
    detail: str


def _compare_canonical(actual: Any, expected_canonical: str, volatile: tuple[str, ...]) -> _Comparison:
    actual_canonical = canonicalize(scrub(actual, volatile))
    if actual_canonical == expected_canonical:
        return _Comparison(passed=True, detail="canonical match")
    return _Comparison(
        passed=False,
        detail=f"expected {expected_canonical}\n       actual {actual_canonical}",
    )


# ── per-op step execution (dispatch dict, no if-chain) ───────────────────────


@dataclass(frozen=True, kw_only=True)
class StepReport:
    index: int
    op: str
    passed: bool
    detail: str


async def _run_list_step(step: ListStep, session: AdapterSession, fixture: Fixture, index: int) -> StepReport:
    actual = await session.list(step.event)
    cmp = _compare_canonical(actual, step.expect.canonical, fixture.volatile)
    return StepReport(index=index, op=step.op, passed=cmp.passed, detail=cmp.detail)


async def _run_invoke_step(step: InvokeStep, session: AdapterSession, fixture: Fixture, index: int) -> StepReport:
    outcome = await session.invoke(step.params)
    if isinstance(step.expect, ErrorExpectation):
        passed = not outcome.ok and outcome.error_code == step.expect.error_code
        detail = (
            f"error {step.expect.error_code} as required"
            if passed
            else f"expected error {step.expect.error_code}, got "
            + ("a result" if outcome.ok else str(outcome.error_code))
        )
        return StepReport(index=index, op=step.op, passed=passed, detail=detail)
    if not outcome.ok:
        return StepReport(
            index=index,
            op=step.op,
            passed=False,
            detail=f"expected a result, got error {outcome.error_code}",
        )
    cmp = _compare_canonical(outcome.result, step.expect.canonical, fixture.volatile)
    return StepReport(index=index, op=step.op, passed=cmp.passed, detail=cmp.detail)


def _check_decision(expect: DecisionExpectation, outcome: ChainOutcome, volatile: tuple[str, ...]) -> _Comparison:
    if outcome.decision != expect.decision:
        return _Comparison(
            passed=False,
            detail=f"expected decision '{expect.decision}', got '{outcome.decision}'",
        )
    final_canonical = canonicalize(scrub(outcome.final_payload, volatile))
    if expect.final_payload_canonical is not None and final_canonical != expect.final_payload_canonical:
        return _Comparison(
            passed=False,
            detail=(
                "final payload mismatch:\n"
                f"       expected {expect.final_payload_canonical}\n"
                f"       actual {final_canonical}"
            ),
        )
    for forbidden in expect.forbids:
        if forbidden in final_canonical:
            return _Comparison(passed=False, detail=f"forbidden content present: {forbidden}")
    for required in expect.requires:
        if required not in final_canonical:
            return _Comparison(passed=False, detail=f"required content absent: {required}")
    return _Comparison(passed=True, detail=f"decision '{expect.decision}' as required")


async def _run_chain_step(step: ChainStep, session: AdapterSession, fixture: Fixture, index: int) -> StepReport:
    outcome = await session.chain(step.event, step.phase, step.payload, step.session_id)
    cmp = _check_decision(step.expect, outcome, fixture.volatile)
    return StepReport(index=index, op=step.op, passed=cmp.passed, detail=cmp.detail)


_StepRunner = Callable[[Any, AdapterSession, Fixture, int], Awaitable[StepReport]]

_RUN_BY_OP: dict[str, _StepRunner] = {
    "list": _run_list_step,
    "invoke": _run_invoke_step,
    "chain": _run_chain_step,
}


# ── fixture + suite execution ────────────────────────────────────────────────


@dataclass(frozen=True, kw_only=True)
class FixtureReport:
    id: str
    requirement: str
    passed: bool
    steps: tuple[StepReport, ...]


@dataclass(frozen=True, kw_only=True)
class ComplianceReport:
    adapter: str
    total: int
    passed: int
    #: passed / total as a percentage, rounded half-up to one decimal place —
    #: the same rounding as the TS reference (`Math.round(x * 1000) / 10`).
    compliance_percent: float
    fixtures: tuple[FixtureReport, ...]


def _round_half_up(value: float) -> int:
    """JS ``Math.round``: half-up toward positive infinity (Python's built-in
    ``round`` is banker's rounding and disagrees on exact .5 ties)."""
    return math.floor(value + 0.5)


def compliance_percent(passed: int, total: int) -> float:
    return 100.0 if total == 0 else _round_half_up((passed / total) * 1000) / 10


async def run_fixture(fixture: Fixture, adapter: Adapter) -> FixtureReport:
    steps: list[StepReport] = []
    async with adapter.create_session(fixture.interceptors) as session:
        for index, step in enumerate(fixture.steps):
            try:
                steps.append(await _RUN_BY_OP[step.op](step, session, fixture, index))
            except Exception as err:  # an adapter crash is a step failure, not a runner crash
                steps.append(
                    StepReport(index=index, op=step.op, passed=False, detail=f"adapter raised: {err}")
                )
    return FixtureReport(
        id=fixture.id,
        requirement=fixture.requirement,
        passed=all(s.passed for s in steps),
        steps=tuple(steps),
    )


async def run_suite(fixtures: tuple[Fixture, ...], adapter: Adapter) -> ComplianceReport:
    reports = [await run_fixture(fixture, adapter) for fixture in fixtures]
    passed = sum(1 for r in reports if r.passed)
    return ComplianceReport(
        adapter=adapter.name,
        total=len(reports),
        passed=passed,
        compliance_percent=compliance_percent(passed, len(reports)),
        fixtures=tuple(reports),
    )


def _js_number(value: float) -> str:
    """Format like JS ``String(n)``: integral floats print without ``.0``."""
    return str(int(value)) if value == int(value) else str(value)


def format_report(report: ComplianceReport) -> str:
    """Human-readable report, line-compatible with the TS ``formatReport``."""
    lines = []
    for f in report.fixtures:
        mark = "PASS" if f.passed else "FAIL"
        failures = "".join(
            f"\n       step {s.index} ({s.op}): {s.detail}" for s in f.steps if not s.passed
        )
        lines.append(f"  {mark}  {f.id} [{f.requirement}]{failures}")
    return "\n".join(
        [
            f"MCP interceptors conformance — adapter: {report.adapter}",
            *lines,
            f"  {report.passed}/{report.total} fixtures passed — "
            f"compliance {_js_number(report.compliance_percent)}%",
        ]
    )

File: conformance/src/runner.ts
/**
 * The conformance RUNNER: replays fixtures against an ADAPTER and diffs
 * canonical JSON. The runner knows nothing about any SDK - everything
 * implementation-specific lives behind the ~4-method adapter contract
 * (ADAPTER.md), which is how Python/C#/Go SDKs reuse this suite without
 * re-authoring tests.
 *
 * Comparison model (identical in every language):
 *   1. scrub volatile keys (fixture.volatile) recursively from the ACTUAL;
 *   2. JCS-canonicalize the scrubbed actual;
 *   3. string-compare against the fixture's precomputed `canonical`.
 * Behavior expectations additionally check decision equality and
 * forbids/requires substrings over the canonical final payload.
 */
import { canonicalize } from "./canonical.ts";
import { isErrorExpectation, STEP_OP } from "./fixture-types.ts";
import type {
  ApplyStep,
  DecisionExpectation,
  Fixture,
  FixtureInterceptor,
  FixtureStep,
  InvokeStep,
  ListStep,
} from "./fixture-types.ts";

// ── the adapter contract (see ADAPTER.md for the prose version) ──────────────

export type InvokeOutcome =
  | { readonly ok: true; readonly result: unknown }
  | { readonly ok: false; readonly errorCode: number };

export interface ChainOutcome {
  /** `allow` iff the chain completed; `deny` iff it was blocked/aborted. */
  readonly decision: "allow" | "deny";
  /** The payload after all applied mutations (meaningful when allowed). */
  readonly finalPayload: unknown;
}

export interface AdapterSession {
  /** Wire JSON of `interceptors/list` (event filter or null). */
  readonly list: (event: string | null) => Promise<unknown>;
  /** Wire JSON of `interceptor/invoke`, or the JSON-RPC error code. */
  readonly invoke: (params: {
    readonly name: string;
    readonly event: string;
    readonly phase: "request" | "response";
    readonly payload: unknown;
  }) => Promise<InvokeOutcome>;
  /** One chain execution across ALL of the session's interceptors. */
  readonly chain: (params: {
    readonly event: string;
    readonly phase: "request" | "response";
    readonly payload: unknown;
    readonly sessionId: string | null;
  }) => Promise<ChainOutcome>;
}

export interface Adapter {
  readonly name: string;
  /** Fresh implementation state with these interceptors registered. */
  readonly createSession: (
    interceptors: readonly FixtureInterceptor[],
  ) => Promise<AdapterSession>;
}

// ── scrubbing + canonical comparison ─────────────────────────────────────────

/** Recursively drop volatile keys (timings, impl-specific info) from actuals. */
export function scrub(value: unknown, volatile: readonly string[]): unknown {
  if (Array.isArray(value)) return value.map((v) => scrub(v, volatile));
  if (typeof value === "object" && value !== null) {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .filter(([k]) => !volatile.includes(k))
        .map(([k, v]) => [k, scrub(v, volatile)]),
    );
  }
  return value;
}

interface Comparison {
  readonly passed: boolean;
  readonly detail: string;
}

function compareCanonical(
  actual: unknown,
  expectedCanonical: string,
  volatile: readonly string[],
): Comparison {
  const actualCanonical = canonicalize(scrub(actual, volatile));
  return actualCanonical === expectedCanonical
    ? { passed: true, detail: "canonical match" }
    : {
        passed: false,
        detail: `expected ${expectedCanonical}\n       actual ${actualCanonical}`,
      };
}

// ── per-op step execution (RULE 2: dispatch table, no switch) ────────────────

export interface StepReport {
  readonly index: number;
  readonly op: FixtureStep["op"];
  readonly passed: boolean;
  readonly detail: string;
}

type StepRunner = (
  step: FixtureStep,
  session: AdapterSession,
  fixture: Fixture,
  index: number,
) => Promise<StepReport>;

async function runListStep(
  step: ListStep,
  session: AdapterSession,
  fixture: Fixture,
  index: number,
): Promise<StepReport> {
  const actual = await session.list(step.event);
  const expected = step.expect.canonical ?? canonicalize(step.expect.result);
  const cmp = compareCanonical(actual, expected, fixture.volatile);
  return { index, op: step.op, ...cmp };
}

async function runInvokeStep(
  step: InvokeStep,
  session: AdapterSession,
  fixture: Fixture,
  index: number,
): Promise<StepReport> {
  const outcome = await session.invoke(step.params);
  if (isErrorExpectation(step.expect)) {
    const passed = !outcome.ok && outcome.errorCode === step.expect.errorCode;
    return {
      index,
      op: step.op,
      passed,
      detail: passed
        ? `error ${String(step.expect.errorCode)} as required`
        : `expected error ${String(step.expect.errorCode)}, got ${
            outcome.ok ? "a result" : String(outcome.errorCode)
          }`,
    };
  }
  if (!outcome.ok) {
    return {
      index,
      op: step.op,
      passed: false,
      detail: `expected a result, got error ${String(outcome.errorCode)}`,
    };
  }
  const expected = step.expect.canonical ?? canonicalize(step.expect.result);
  const cmp = compareCanonical(outcome.result, expected, fixture.volatile);
  return { index, op: step.op, ...cmp };
}

function checkDecision(
  expect: DecisionExpectation,
  outcome: ChainOutcome,
  volatile: readonly string[],
): Comparison {
  if (outcome.decision !== expect.decision) {
    return {
      passed: false,
      detail: `expected decision '${expect.decision}', got '${outcome.decision}'`,
    };
  }
  const finalCanonical =
    outcome.finalPayload === undefined
      ? "null"
      : canonicalize(scrub(outcome.finalPayload, volatile));
  if (expect.finalPayload !== undefined) {
    const expected =
      expect.finalPayloadCanonical ?? canonicalize(expect.finalPayload);
    if (finalCanonical !== expected) {
      return {
        passed: false,
        detail: `final payload mismatch:\n       expected ${expected}\n       actual ${finalCanonical}`,
      };
    }
  }
  const leaked = (expect.forbids ?? []).find((s) => finalCanonical.includes(s));
  if (leaked !== undefined) {
    return { passed: false, detail: `forbidden content present: ${leaked}` };
  }
  const missing = (expect.requires ?? []).find((s) => !finalCanonical.includes(s));
  if (missing !== undefined) {
    return { passed: false, detail: `required content absent: ${missing}` };
  }
  return { passed: true, detail: `decision '${expect.decision}' as required` };
}

async function runApplyStep(
  step: ApplyStep,
  session: AdapterSession,
  fixture: Fixture,
  index: number,
): Promise<StepReport> {
  const outcome = await session.chain({
    event: step.event,
    phase: step.phase,
    payload: step.payload,
    sessionId: step.context?.sessionId ?? null,
  });
  const cmp = checkDecision(step.expect, outcome, fixture.volatile);
  return { index, op: step.op, ...cmp };
}

const RUN_BY_OP: Record<FixtureStep["op"], StepRunner> = {
  [STEP_OP.List]: (step, ...rest) => runListStep(step as ListStep, ...rest),
  [STEP_OP.Invoke]: (step, ...rest) => runInvokeStep(step as InvokeStep, ...rest),
  [STEP_OP.Apply]: (step, ...rest) => runApplyStep(step as ApplyStep, ...rest),
};

// ── fixture + suite execution ────────────────────────────────────────────────

export interface FixtureReport {
  readonly id: string;
  readonly requirement: string;
  readonly passed: boolean;
  readonly steps: readonly StepReport[];
}

export interface ComplianceReport {
  readonly adapter: string;
  readonly total: number;
  readonly passed: number;
  /** passed / total, rounded to one decimal place. */
  readonly compliancePercent: number;
  readonly fixtures: readonly FixtureReport[];
}

export async function runFixture(
  fixture: Fixture,
  adapter: Adapter,
): Promise<FixtureReport> {
  const session = await adapter.createSession(fixture.interceptors);
  const steps: StepReport[] = [];
  for (const [index, step] of fixture.steps.entries()) {
    try {
      steps.push(await RUN_BY_OP[step.op](step, session, fixture, index));
    } catch (err) {
      steps.push({
        index,
        op: step.op,
        passed: false,
        detail: `adapter threw: ${err instanceof Error ? err.message : String(err)}`,
      });
    }
  }
  return {
    id: fixture.id,
    requirement: fixture.requirement,
    passed: steps.every((s) => s.passed),
    steps,
  };
}

export async function runSuite(
  fixtures: readonly Fixture[],
  adapter: Adapter,
): Promise<ComplianceReport> {
  const reports: FixtureReport[] = [];
  for (const fixture of fixtures) {
    reports.push(await runFixture(fixture, adapter));
  }
  const passed = reports.filter((r) => r.passed).length;
  return {
    adapter: adapter.name,
    total: reports.length,
    passed,
    compliancePercent:
      reports.length === 0 ? 100 : Math.round((passed / reports.length) * 1000) / 10,
    fixtures: reports,
  };
}

/** Human-readable report: per-fixture status and the compliance percentage. */
export function formatReport(report: ComplianceReport): string {
  const lines = report.fixtures.map((f) => {
    const mark = f.passed ? "PASS" : "FAIL";
    const failures = f.steps
      .filter((s) => !s.passed)
      .map((s) => `\n       step ${String(s.index)} (${s.op}): ${s.detail}`)
      .join("");
    return `  ${mark}  ${f.id} [${f.requirement}]${failures}`;
  });
  return [
    `MCP interceptors conformance - adapter: ${report.adapter}`,
    ...lines,
    `  ${String(report.passed)}/${String(report.total)} fixtures passed - compliance ${String(report.compliancePercent)}%`,
  ].join("\n");
}
